### ISSUE TITLE
切片操作问题

### ISSUE THREAD

**Issue by lcylcy** [Oneflow-Inc/oneflow#6360]
···
#或操作
def p1():
    a = flow.randn(4,4)
    print(a)

    s1 = a>0.3
    s2 = a<=-0.2
    print(s1)
    print(s2)
    print(s1 | s2)

```
#索引操作
def p2():
    a = np.random.rand(4,4)

    oi = flow.Tensor(a)
    ms = oi>0.5
    
    print(oi[ms].shape)

    oi[ms] = 0
    print(oi)

```
#torch和oneflow int32索引不一致
def p3():
    n = np.random.rand(3,4)

    ti = torch.tensor(n)
    mask = ti>0.2
    ti[mask] = 0
    print(ti[mask].shape)

    oi = flow.Tensor(n)
    mask = oi>0.2
    mask = mask.to(flow.int32)
    oi[mask] = 0
    print(oi[mask].shape)

**Comment by lcylcy** [comment]
```
#索引操作
def p2():
    a = np.random.rand(4,4)
    oi = flow.Tensor(a)
    ms = oi>0.5
    print(oi[ms].shape)
    oi[ms] = 0
    print(oi)

**Comment by lcylcy** [comment]
```
#torch和oneflow int32索引不一致
def p3():
    n = np.random.rand(3,4)
    ti = torch.tensor(n)
    mask = ti>0.2
    ti[mask] = 0
    print(ti[mask].shape)
    oi = flow.Tensor(n)
    mask = oi>0.2
    mask = mask.to(flow.int32)
    oi[mask] = 0
    print(oi[mask].shape)

**Comment by hjchen2** [comment]
> #索引操作
> def p2():
>     a = np.random.rand(4,4)
>     oi = flow.Tensor(a)
>     ms = oi>0.5
>     print(oi[ms].shape)
>     oi[ms] = 0
>     print(oi)

这里有两个问题，一个是bool tensor的支持，一个是scatter update，前一个问题需要将bool tensor解析成正常的index再切片，后一个问题比较简单，新加一个op就行了。

**Comment by hjchen2** [comment]
> > #索引操作
> > def p2():
> > a = np.random.rand(4,4)
> > oi = flow.Tensor(a)
> > ms = oi>0.5
> > print(oi[ms].shape)
> > oi[ms] = 0
> > print(oi)
> 
> 这里有两个问题，一个是bool tensor的支持，一个是scatter update，前一个问题需要将bool tensor解析成正常的index再切片，后一个问题比较简单，新加一个op就行了。

scatter update的问题 @PragmaTwice 明阳看一下，和scatter_nd的实现逻辑差不多。

**Comment by hjchen2** [comment]
> #torch和oneflow int32索引不一致
> def p3():
>     n = np.random.rand(3,4)
>     ti = torch.tensor(n)
>     mask = ti>0.2
>     ti[mask] = 0
>     print(ti[mask].shape)
>     oi = flow.Tensor(n)
>     mask = oi>0.2
>     mask = mask.to(flow.int32)
>     oi[mask] = 0
>     print(oi[mask].shape)

不是Int32索引不一致，而是你pytorch用的是bool的索引，oneflow你用的是int32索引，本身pytorch bool索引的逻辑和int类型的就不大一样，结果肯定不一样了。

**Comment by hjchen2** [comment]
> #torch和oneflow int32索引不一致
> def p3():
>     n = np.random.rand(3,4)
>     ti = torch.tensor(n)
>     mask = ti>0.2
>     ti[mask] = 0
>     print(ti[mask].shape)
>     oi = flow.Tensor(n)
>     mask = oi>0.2
>     mask = mask.to(flow.int32)
>     oi[mask] = 0
>     print(oi[mask].shape)

mask select已经支持了，clone最新的master代码重新编译一下再试试。

**Comment by lcylcy** [comment]
ok




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Houjiang ***@***.***&gt;; 
发送时间: 2021年9月24日(星期五) 中午12:02
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [Oneflow-Inc/oneflow] 切片操作问题 (#6360)





  
#torch和oneflow int32索引不一致
 def p3():
 n = np.random.rand(3,4)
 ti = torch.tensor(n)
 mask = ti&gt;0.2
 ti[mask] = 0
 print(ti[mask].shape)
 oi = flow.Tensor(n)
 mask = oi&gt;0.2
 mask = mask.to(flow.int32)
 oi[mask] = 0
 print(oi[mask].shape)
  
mask select已经支持了，clone最新的master代码重新编译一下再试试。
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
